File: ligature-xodus/shared/src/main/scala/dev/ligature/xodus/XodusLigature.scala
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.xodus

import dev.ligature.{
  Dataset,
  Identifier,
  Ligature,
  LigatureError,
  QueryTx,
  Statement,
  Value,
  WriteTx
}
import cats.effect.IO
import fs2.Stream

import scala.collection.immutable.TreeMap
import scala.collection.mutable.{ArrayBuffer, ListBuffer}
import cats.effect.kernel.Resource
import cats.effect.kernel.Ref
import jetbrains.exodus.ByteIterable

import java.io.File
import jetbrains.exodus.env.{
  EnvironmentConfig,
  Environments,
  ReadonlyTransaction,
  Store,
  StoreConfig,
  Transaction,
  TransactionalComputable,
  TransactionalExecutable
}
import jetbrains.exodus.bindings.{LongBinding, StringBinding}

import scala.jdk.CollectionConverters.*

trait XodusOperations {
  def openStore(tx: Transaction, store: LigatureStore): Store
  def nextID(tx: Transaction): ByteIterable
}

enum LigatureStore(val storeName: String):
  case CountersStore extends LigatureStore("Counters")
  case DatasetToIdStore extends LigatureStore("DatasetToID")
  case IdToDatasetStore extends LigatureStore("IDToDataset")
  case EAVStore extends LigatureStore("EAV")
  case EVAStore extends LigatureStore("EVA")
  case AEVStore extends LigatureStore("AEV")
  case AVEStore extends LigatureStore("AVE")
  case VEAStore extends LigatureStore("VEA")
  case VAEStore extends LigatureStore("VAE")
  case IdentifierToIdStore extends LigatureStore("IdentifierToID")
  case IdToIdentifierStore extends LigatureStore("IDToIdentifier")
  case StringToIdStore extends LigatureStore("StringToID")
  case IdToStringStore extends LigatureStore("IDToString")
  case BytesToIdStore extends LigatureStore("BytesToID")
  case IdToBytesStore extends LigatureStore("IDToBytes")

enum LigatureValueType(val id: Byte):
  case Identifier extends LigatureValueType(0)
  case Integer extends LigatureValueType(1)
  case String extends LigatureValueType(2)
  case Bytes extends LigatureValueType(3)

final class XodusLigature(dbDirectory: File) extends Ligature with XodusOperations {
  private val environment = Environments.newInstance(dbDirectory, new EnvironmentConfig)
  setupStores()

  /** This method is ran once at the start to make sure all Stores exist.
    * This is done so that Stores can be opened in readonly mode.
    *  It also checks and initializes Counters
    * TODO: This method could probably check if the Environment exists and check the status of the Environment first.
    */
  private def setupStores(): Unit = {
    // create all Stores
    val createStoresTC: TransactionalComputable[Unit] = tx =>
      LigatureStore.values.foreach(openStore(tx, _))
    environment.computeInTransaction(createStoresTC)
    // set up counters -- for now it's just a single counter
    val checkCountersTC: TransactionalComputable[Unit] = tx => {
      val counterStore = openStore(tx, LigatureStore.CountersStore)
      val currentCount = counterStore.get(tx, StringBinding.stringToEntry("counter"))
      if (currentCount == null) {
        counterStore.put(tx, StringBinding.stringToEntry("counter"), LongBinding.longToEntry(0))
      }
    }
    environment.computeInTransaction(checkCountersTC)
  }

  /** Used to uniformly open Stores.
    * This is mainly so that StoreConfigs can be consistent.
    * Right now all stores use the same config but that is likely to change.
    */
  override def openStore(tx: Transaction, store: LigatureStore): Store =
    environment.openStore(store.storeName, StoreConfig.WITHOUT_DUPLICATES, tx)

  /** Computes the next ID.
    * Right now IDs are shared between all Stores but that might change.
    * Also, right now IDs are positive Long values and that might also change.
    */
  override def nextID(tx: Transaction): ByteIterable = {
    val counterStore = openStore(tx, LigatureStore.CountersStore)
    val nextCount =
      LongBinding.entryToLong(counterStore.get(tx, StringBinding.stringToEntry("counter"))) + 1L
    counterStore.put(tx, StringBinding.stringToEntry("counter"), LongBinding.longToEntry(nextCount))
    LongBinding.longToEntry(nextCount)
  }

  /** Returns all Datasets in a Ligature instance. */
  override def allDatasets(): Stream[IO, Dataset] = Stream.emits {
    val tc: TransactionalComputable[List[String]] = tx => {
      val datasetToIdStore = openStore(tx, LigatureStore.DatasetToIdStore)
      val datasetsCursor = datasetToIdStore.openCursor(tx)
      val datasets: ListBuffer[String] = ListBuffer()
      while (datasetsCursor.getNext)
        datasets.append(StringBinding.entryToString(datasetsCursor.getKey))
      datasetsCursor.close() // TODO should use bracket
      datasets.toList
    }
    environment.computeInReadonlyTransaction(tc).map(Dataset.fromString(_).getOrElse(???))
  }

  /** Check if a given Dataset exists. */
  override def datasetExists(dataset: Dataset): IO[Boolean] = IO {
    val tc: TransactionalComputable[Boolean] = tx => {
      val datasetToIdStore = openStore(tx, LigatureStore.DatasetToIdStore)
      val res = datasetToIdStore.get(tx, StringBinding.stringToEntry(dataset.name))
      res != null
    }
    environment.computeInReadonlyTransaction(tc)
  }

  /** Returns all Datasets in a Ligature instance that start with the given
    * prefix.
    */
  override def matchDatasetsPrefix(prefix: String): Stream[IO, Dataset] = Stream.emits {
    val tc: TransactionalComputable[Seq[Dataset]] = tx => {
      val results: ArrayBuffer[Dataset] = ArrayBuffer()
      val cursor = openStore(tx, LigatureStore.DatasetToIdStore).openCursor(tx)
      val prefixBytes = StringBinding.stringToEntry(prefix)
      cursor.getSearchKeyRange(prefixBytes)
      var r = cursor.getKey
      var continue = true
      while (continue && r != null && StringBinding.entryToString(r).startsWith(prefix)) {
        results.append(Dataset.fromString(StringBinding.entryToString(r)).getOrElse(???))
        continue = cursor.getNext
        r = cursor.getKey
      }
      results.toSeq
    }
    environment.computeInReadonlyTransaction(tc)
  }

  /** Returns all Datasets in a Ligature instance that are in a given range
    * (inclusive, exclusive].
    */
  override def matchDatasetsRange(start: String, end: String): Stream[IO, Dataset] = Stream.emits {
    val tc: TransactionalComputable[Seq[Dataset]] = tx => {
      val results: ArrayBuffer[Dataset] = ArrayBuffer()
      val cursor = openStore(tx, LigatureStore.DatasetToIdStore).openCursor(tx)
      val prefixBytes = StringBinding.stringToEntry(start)
      cursor.getSearchKeyRange(prefixBytes)
      var r = cursor.getKey
      var continue = true
      while (continue && r != null && stringInRange(StringBinding.entryToString(r), start, end)) {
        results.append(Dataset.fromString(StringBinding.entryToString(r)).getOrElse(???))
        continue = cursor.getNext
        r = cursor.getKey
      }
      results.toSeq
    }
    environment.computeInReadonlyTransaction(tc)
  }

  private def stringInRange(toTest: String, start: String, end: String): Boolean =
    toTest.compareTo(start) >= 0 && toTest.compareTo(end) < 0

  /** Creates a dataset with the given name. TODO should probably return its own
    * error type { InvalidDataset, DatasetExists, CouldNotCreateDataset }
    */
  override def createDataset(dataset: Dataset): IO[Unit] = IO {
    val tc: TransactionalExecutable = tx => {
      val nameEntry = StringBinding.stringToEntry(dataset.name)
      val datasetToIdStore = openStore(tx, LigatureStore.DatasetToIdStore)
      val datasetExists = datasetToIdStore.get(tx, nameEntry)
      if (datasetExists == null) {
        val nextId = nextID(tx)
        val idToDatasetStore = openStore(tx, LigatureStore.IdToDatasetStore)
        datasetToIdStore.put(tx, nameEntry, nextId)
        idToDatasetStore.putRight(tx, nextId, nameEntry)
        tx.commit()
      } else {
        ()
      }
    }
    environment.executeInTransaction(tc)
  }

  /** Deletes a dataset with the given name. TODO should probably return its own
    * error type { InvalidDataset, CouldNotDeleteDataset }
    */
  override def deleteDataset(dataset: Dataset): IO[Unit] = IO {
    val tc: TransactionalExecutable = tx => {
      val nameEntry = StringBinding.stringToEntry(dataset.name)
      val datasetToIdStore = openStore(tx, LigatureStore.DatasetToIdStore)
      val datasetId = datasetToIdStore.get(tx, nameEntry)
      if (datasetId != null) {
        val idToDatasetStore = openStore(tx, LigatureStore.IdToDatasetStore)
        datasetToIdStore.delete(tx, nameEntry)
        idToDatasetStore.delete(tx, datasetId)
        // TODO remove all statements
        tx.commit()
      } else {
        ()
      }
    }
    environment.executeInTransaction(tc)
  }

  /** Initializes a QueryTx TODO should probably return its own error type
    * CouldNotInitializeQueryTx
    */
  override def query[T](dataset: Dataset)(fn: QueryTx => IO[T]): IO[T] =
    IO { //TODO rewrite with TransactionComputable
      environment.beginReadonlyTransaction()
    }.bracket { tx => IO.defer {
      val queryTx = XodusQueryTx(tx, this, dataset)
      fn(queryTx)
    }} { tx => IO.defer {
      if (!tx.isFinished) {
        tx.abort()
      }
      IO.unit
    }}

  /** Initializes a WriteTx TODO should probably return its own error type
    * CouldNotInitializeWriteTx
    */
  override def write(dataset: Dataset)(fn: WriteTx => IO[Unit]): IO[Unit] = IO {
    val te: TransactionalExecutable = tx => {
      val writeTx = XodusWriteTx(tx, this, dataset)
      fn(writeTx)
    }
    environment.executeInTransaction(te)
  }

  override def close(): IO[Unit] = IO {
    environment.close()
    ()
  }
}

File: ligature-xodus/src/jvmMain/kotlin/dev/ligature/xodus/XodusLigature.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at https://mozilla.org/MPL/2.0/. */

package dev.ligature.xodus

import arrow.core.Either.Left
import arrow.core.Either.Right
import dev.ligature.Dataset
import dev.ligature.Identifier
import dev.ligature.Ligature
import dev.ligature.LigatureError
import dev.ligature.QueryTx
import dev.ligature.Statement
import dev.ligature.Value
import dev.ligature.WriteTx

import jetbrains.exodus.ByteIterable

import java.io.File
import jetbrains.exodus.env.EnvironmentConfig
import jetbrains.exodus.env.Environments
import jetbrains.exodus.env.ReadonlyTransaction
import jetbrains.exodus.env.Store
import jetbrains.exodus.env.StoreConfig
import jetbrains.exodus.env.Transaction
import jetbrains.exodus.env.TransactionalComputable
import jetbrains.exodus.env.TransactionalExecutable

import jetbrains.exodus.bindings.LongBinding
import jetbrains.exodus.bindings.StringBinding
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

interface XodusOperations {
  fun openStore(tx: Transaction, store: LigatureStore): Store
  fun nextID(tx: Transaction): ByteIterable
}

enum class LigatureStore(val storeName: String) {
  CountersStore("Counters"),
  DatasetToIdStore("DatasetToID"),
  IdToDatasetStore("IDToDataset"),
  EAVStore("EAV"),
  EVAStore("EVA"),
  AEVStore("AEV"),
  AVEStore("AVE"),
  VEAStore("VEA"),
  VAEStore("VAE"),
  IdentifierToIdStore("IdentifierToID"),
  IdToIdentifierStore("IDToIdentifier"),
  StringToIdStore("StringToID"),
  IdToStringStore("IDToString"),
  BytesToIdStore("BytesToID"),
  IdToBytesStore("IDToBytes")
}

enum class LigatureValueType(val id: Byte) {
  Identifier(0),
  Integer(1),
  String(2),
  Bytes(3);

  companion object {
    fun getValueType(id: Byte): LigatureValueType = when (id) {
      LigatureValueType.Identifier.id -> LigatureValueType.Identifier
      LigatureValueType.Integer.id -> LigatureValueType.Integer
      LigatureValueType.String.id -> LigatureValueType.String
      LigatureValueType.Bytes.id -> LigatureValueType.Bytes
      else -> TODO("handle error case, invalid type")
    }
  }
}

class XodusLigature(private val dbDirectory: File): Ligature, XodusOperations {
  private val environment = Environments.newInstance(dbDirectory, EnvironmentConfig())

  init {
    setupStores()
  }

  /** This method is ran once at the start to make sure all Stores exist.
    * This is done so that Stores can be opened in readonly mode.
    *  It also checks and initializes Counters
    * TODO: This method could probably check if the Environment exists and check the status of the Environment first.
    */
  private fun setupStores(): Unit {
    // create all Stores
    val createStoresTC: TransactionalComputable<Unit> = TransactionalComputable<Unit> { tx ->
      LigatureStore.values().forEach { openStore(tx, it) }
    }
    environment.computeInTransaction(createStoresTC)
    // set up counters -- for now it's just a single counter
    val checkCountersTC: TransactionalComputable<Unit> = TransactionalComputable<Unit> { tx ->
      val counterStore = openStore(tx, LigatureStore.CountersStore)
      val currentCount = counterStore.get(tx, StringBinding.stringToEntry("counter"))
      if (currentCount == null) {
        counterStore.put(tx, StringBinding.stringToEntry("counter"), LongBinding.longToEntry(0))
      }
    }
    environment.computeInTransaction(checkCountersTC)
  }

  /** Used to uniformly open Stores.
    * This is mainly so that StoreConfigs can be consistent.
    * Right now all stores use the same config but that is likely to change.
    */
  override fun openStore(tx: Transaction, store: LigatureStore): Store =
    environment.openStore(store.storeName, StoreConfig.WITHOUT_DUPLICATES, tx)

  /** Computes the next ID.
    * Right now IDs are shared between all Stores but that might change.
    * Also, right now IDs are positive Long values and that might also change.
    */
  override fun nextID(tx: Transaction): ByteIterable {
    val counterStore = openStore(tx, LigatureStore.CountersStore)
    val nextCount =
      LongBinding.entryToLong(counterStore.get(tx, StringBinding.stringToEntry("counter"))!!) + 1L
    counterStore.put(tx, StringBinding.stringToEntry("counter"), LongBinding.longToEntry(nextCount))
    return LongBinding.longToEntry(nextCount)
  }

  /** Returns all Datasets in a Ligature instance. */
  override fun allDatasets(): Flow<Dataset> = flow {
    val tc = TransactionalComputable<List<String>> { tx ->
      val datasetToIdStore = openStore(tx, LigatureStore.DatasetToIdStore)
      val datasetsCursor = datasetToIdStore.openCursor(tx)
      val datasets: MutableList<String> = mutableListOf()
      while (datasetsCursor.next)
        datasets.add(StringBinding.entryToString(datasetsCursor.key))
      datasetsCursor.close() // TODO should use bracket
      datasets
    }
    environment.computeInReadonlyTransaction(tc).map { Dataset.create(it) }.forEach {
      when (it) {
        is Right -> emit(it.value)
        is Left  -> TODO()
      }
    }
  }

  private fun fetchDatasetID(dataset: Dataset): ByteIterable? {
    val tc = TransactionalComputable<ByteIterable?> { tx ->
      val datasetToIdStore = openStore(tx, LigatureStore.DatasetToIdStore)
      datasetToIdStore.get(tx, StringBinding.stringToEntry(dataset.name))
    }
    return environment.computeInReadonlyTransaction(tc)
  }

  /** Check if a given Dataset exists. */
  override suspend fun datasetExists(dataset: Dataset): Boolean =
    fetchDatasetID(dataset) != null

  /** Returns all Datasets in a Ligature instance that start with the given
    * prefix.
    */
  override fun matchDatasetsPrefix(prefix: String): Flow<Dataset> = flow {
    val tc = TransactionalComputable<List<Dataset>> { tx ->
      val results = mutableListOf<Dataset>()
      val cursor = openStore(tx, LigatureStore.DatasetToIdStore).openCursor(tx)
      val prefixBytes = StringBinding.stringToEntry(prefix)
      cursor.getSearchKeyRange(prefixBytes)
      var r: ByteIterable? = cursor.key
      var `continue` = true
      while (`continue` && r != null && StringBinding.entryToString(r).startsWith(prefix)) {
        when (val ds = Dataset.create(StringBinding.entryToString(r))) {
          is Right -> {
            results.add(ds.value)
            `continue` = cursor.next
            r = cursor.key
          }
          is Left -> {
            TODO()
          }
        }
      }
      results
    }
    environment.computeInReadonlyTransaction(tc).forEach { emit(it) }
  }

  /** Returns all Datasets in a Ligature instance that are in a given range
    * (inclusive, exclusive].
    */
  override fun matchDatasetsRange(start: String, end: String): Flow<Dataset> = flow {
    val tc = TransactionalComputable<List<Dataset>> { tx ->
      val results = mutableListOf<Dataset>()
      val cursor = openStore(tx, LigatureStore.DatasetToIdStore).openCursor(tx)
      val prefixBytes = StringBinding.stringToEntry(start)
      cursor.getSearchKeyRange(prefixBytes)
      var r: ByteIterable? = cursor.key
      var `continue` = true
      while (`continue` && r != null && stringInRange(StringBinding.entryToString(r), start, end)) {
        when (val ds = Dataset.create(StringBinding.entryToString(r))) {
          is Right -> {
            results.add(ds.value)
            `continue` = cursor.next
            r = cursor.key
          }
          is Left -> {
            TODO()
          }
        }
      }
      results
    }
    environment.computeInReadonlyTransaction(tc)
  }

  private fun stringInRange(toTest: String, start: String, end: String): Boolean =
    toTest >= start && toTest < end

  /** Creates a dataset with the given name. TODO should probably return its own
    * error type { InvalidDataset, DatasetExists, CouldNotCreateDataset }
    */
  override suspend fun createDataset(dataset: Dataset) {
    val tc = TransactionalExecutable { tx ->
      val nameEntry = StringBinding.stringToEntry(dataset.name)
      val datasetToIdStore = openStore(tx, LigatureStore.DatasetToIdStore)
      val datasetExists = datasetToIdStore.get(tx, nameEntry)
      if (datasetExists == null) {
        val nextId = nextID(tx)
        val idToDatasetStore = openStore(tx, LigatureStore.IdToDatasetStore)
        datasetToIdStore.put(tx, nameEntry, nextId)
        idToDatasetStore.putRight(tx, nextId, nameEntry)
        tx.commit()
      }
    }
    environment.executeInTransaction(tc)
  }

  /** Deletes a dataset with the given name. TODO should probably return its own
    * error type { InvalidDataset, CouldNotDeleteDataset }
    */
  override suspend fun deleteDataset(dataset: Dataset) {
    val tc = TransactionalExecutable { tx ->
      val nameEntry = StringBinding.stringToEntry(dataset.name)
      val datasetToIdStore = openStore(tx, LigatureStore.DatasetToIdStore)
      val datasetId = datasetToIdStore.get(tx, nameEntry)
      if (datasetId != null) {
        val idToDatasetStore = openStore(tx, LigatureStore.IdToDatasetStore)
        datasetToIdStore.delete(tx, nameEntry)
        idToDatasetStore.delete(tx, datasetId)
        deleteAllStatements(tx, datasetId)
        tx.commit()
      }
    }
    environment.executeInTransaction(tc)
  }

  /** This method iterates through all the index stores and removes all Statements
    * in a given Dataset.
    * TODO this method also needs to clean up all Identifiers, Strings, and Bytes
    */
  private fun deleteAllStatements(tx: Transaction, datasetId: ByteIterable) {
    val stores = listOf(
      LigatureStore.EAVStore,
      LigatureStore.EVAStore,
      LigatureStore.AEVStore,
      LigatureStore.AVEStore,
      LigatureStore.VEAStore,
      LigatureStore.VAEStore,
      LigatureStore.IdentifierToIdStore,
      LigatureStore.IdToIdentifierStore,
      LigatureStore.StringToIdStore,
      LigatureStore.IdToStringStore,
      LigatureStore.BytesToIdStore,
      LigatureStore.IdToBytesStore
    )
    stores.forEach { store ->
      val cursor = openStore(tx, store).openCursor(tx)
      var `continue` = cursor.getSearchKeyRange(datasetId) != null
      while (`continue`) {
        val key = cursor.key
        `continue` = if (key.subIterable(0, 8) == datasetId) {
          cursor.deleteCurrent()
          cursor.next
        } else {
          false
        }
      }
    }
  }

  /** Initializes a QueryTx TODO should probably return its own error type
    * CouldNotInitializeQueryTx
    */
  //TODO probably use arrow's Resource here?
  override suspend fun <T>query(dataset: Dataset, fn: suspend (QueryTx) -> T): T {
    val tx = environment.beginReadonlyTransaction()
    val res = when (val datasetId = fetchDatasetID(dataset)) {
      null -> TODO()
      else -> {
        val queryTx = XodusQueryTx(tx, this, datasetId)
        fn(queryTx)
      }
    }
    if (!tx.isFinished) {
      tx.abort()
    }
    return res
  }

  /** Initializes a WriteTx TODO should probably return its own error type
    * CouldNotInitializeWriteTx
    */
  //TODO probably use arrow's Resource here?
  override suspend fun <T>write(dataset: Dataset, fn: suspend (WriteTx) -> T): T {
    val tx = environment.beginTransaction()
    val res = when(val datasetId = fetchDatasetID(dataset)) {
      null -> TODO()
      else -> {
        val writeTx = XodusWriteTx(tx, this, datasetId)
        fn(writeTx)
      }
    }
    if (!tx.isFinished) {
      tx.commit()
    }
    return res
  }

  override suspend fun close(): Unit =
    environment.close()
}
